safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,267 @@
+
+            backup_lsn = backup_lsn_range(backup_lsn, commit_lsn, wal_seg_size, &conf, timeline_id).await?;
+            lsn_backed_up.send(backup_lsn)?;
+        }
+    }
+}
+
+pub async fn backup_lsn_range(start_lsn: Lsn, end_lsn: Lsn, wal_seg_size: usize, conf: &SafeKeeperConf, timeline_id: ZTenantTimelineId) -> Result<Lsn> {
+
+    let mut res = start_lsn;
+    for s in get_segments(start_lsn, end_lsn, wal_seg_size) {
+
+        let seg_backup = backup_single_segment(s, wal_seg_size, conf, timeline_id).await;
+
+        // TODO: limit this only to Not Found errors
+        if seg_backup.is_err() && start_lsn.is_valid() {
+            error!("Segment {} not found in timeline {}", s.seg_no, timeline_id)
+        }
+
+        ensure!(start_lsn >= s.start_lsn || start_lsn.is_valid(), "Out of order segment upload detected");
+
+        if res == s.start_lsn {
+            res = s.end_lsn;
+        } else {
+            warn!("Out of order Segment {} upload had been detected for timeline {}. Backup Lsn {}, Segment Start Lsn {}", s.seg_no, timeline_id, res, s.start_lsn)
+        }
+    }
+
+    Ok(res)
+}
+
+async fn backup_single_segment(seg: Segment, wal_seg_size: usize, conf: &SafeKeeperConf, timeline_id: ZTenantTimelineId) -> Result<()> {
+
+    let segment_file_name = seg.file_path(&timeline_id, conf)?;
+    let dest_name = PathBuf::from(format!("{}/{}", timeline_id.tenant_id, timeline_id.timeline_id)).with_file_name(seg.file_name());
+
+    info!("Backup of {:?} requested", segment_file_name);
+    ensure!(seg.size() == wal_seg_size);
+
+    if !segment_file_name.exists() {
+        // TODO: antons return a specific error
+        bail!("Segment file is Missing");
+    } 
+
+    backup_object(conf.remote_storage_config.as_ref(), conf.timeline_dir(&timeline_id), &segment_file_name, seg.size(), &dest_name).await?;
+
+    info!("Backup of {:?} done", segment_file_name);
+
+    Ok(())
+}
+
+pub fn create(conf: &SafeKeeperConf, timeline_id: &ZTenantTimelineId, initial_lsn: Lsn, wal_seg_size: Receiver<u32>, lsn_committed: Receiver<Lsn>, lsn_backed_up: Sender<Lsn>) -> Result<()> {
+    info!("Creating wal backup task for timeline {}, starting with Lsn {}", timeline_id.timeline_id, initial_lsn);
+    let runtime = BACKUP_RUNTIME.get_or_init(|| {
+        let rt_size =  usize::try_from(conf.backup_runtime_threads.unwrap_or(DEFAULT_BACKUP_RUNTIME_SIZE))
+            .expect("Could not get configuration value for backup_runtime_threads");
+
+        info!("initialize backup async runtime with {} threads", rt_size);
+
+        Builder::new_multi_thread()
+            .worker_threads(rt_size)
+            .enable_all()
+            .build()
+            .expect("Couldn't create wal backup runtime")
+    });
+
+    let lease_id = conf.broker_endpoints
+        .as_ref()
+        .map(|_| {
+            let lease = broker::get_lease(conf);
+            runtime.spawn(broker::lease_keep_alive(lease, conf.clone()));
```

Comment:
This lease seems not to be released once timeline is deactivated. We can create WalBackup struct which will "hold" the lease, and release it in the destructor.

Reply:
added ability to cancel WAL backup, but haven't wired it to [de]activation.
spilling this functionality over into https://github.com/neondatabase/neon/issues/1644